neural_speed/models/llama/llama_yarn.cpp

```diff
@@ -0,0 +1,525 @@
+//  Copyright (c) 2023 Intel Corporation
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+#include "models/llama/llama.h"
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cinttypes>
+#include <cstring>
+#include <exception>
+#include <fstream>
+#include <iterator>
+#include <memory>
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "core/data_types.h"
+#include "core/layers/mha_dense.h"
+#include "core/ne.h"
+#include "core/ne_bestla.h"
+#include "core/ne_layers.h"
+#include "models/model_utils/model_config.h"
+#include "models/model_utils/model_files.h"
+#include "models/model_utils/model_types.h"
+#include "models/model_utils/model_utils.h"
+#include "models/model_utils/util.h"
+#include "models/models.h"
+
+// evaluate the transformer
+//
+//   - lctx:      model context
+//   - inputs:    model_input array
+//   - n_input    num of model_input
+//   - n_threads: number of threads to use
+//
+static bool llama_model_eval_internal(model_context* ctx, const model_input* inputs, const int n_input,
+                                      const int n_threads) {
+  model_context& lctx = *ctx;
```

Comment:
this is llama basically, can you reuse the code as much as possible?

Reply:
@airMeng 
Yes, this is llama architecuture based model.
The difference is there is bias in attention layer.

The problem is I did not fina a way to get to weight name of each attention layer.
If we do not plan to change the data structure of layer, we cannot reuse the llama code easily. I suppoe the layer data structure is common to all other models. So it is created as a new model. 

![image](https://github.com/intel/neural-speed/assets/111278656/424f9caf-58b6-47bc-9581-10ce049b4425)

Reply:
An example, just for reference
```
// llama.cpp
#ifdef YARN
#define ATTENTION_BIAS_IDX 2
#else
#define ATTENTION_BIAS_IDX 1
#endif

...model.layers[ii].attn[ATTENTION_BIAS_IDX]
```

in [CMakeLists.txt](https://github.com/intel/neural-speed/blob/main/neural_speed/models/CMakeLists.txt)
```
function(add_model target yarn)
  if(NOT DEFINED yarn)
    set(yarn FALSE)
  endif()
  if (yarn) 
      add_compile_definitions(YARN)
  endif()
  add_library_w_warning(${target} ${ARGN})  # no (gpt) model utils needed
  target_compile_features(${target} PUBLIC cxx_std_11) # don't bump
  set_target_properties(${target} PROPERTIES POSITION_INDEPENDENT_CODE ON)
  target_link_libraries(${target} PUBLIC ne_layers bestla)
endfunction()

add_model(llama TRUE llama/llama.cpp llama/llama_utils.cpp ${MODEL_UTILS_SOURCE})
```

I don't know whether the above works exactly, you can refer to @intellinjun for details

Reply:
@airMeng @intellinjun 
Reuse llama code,Done.
Upload the code. 
Please help to review it, thanks!